sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py

```diff
@@ -478,3 +487,199 @@ def test_splitting_missing_values(X_binned, all_gradients,
+
+        # all categories appear less than CAT_SMOOTH
+        ([0] * 9 + [1] * 8, False, 2),
+
+        # only one category appear more than CAT_SMOOTH
+        ([0] * 12 + [1] * 8, False, 2),
+
+        # missing values + category appear less than CAT_SMOOTH
+        # 9 is missing
+        ([0] * 9 + [1] * 8 + [9] * 4, True, 2),
+
+        # 9 is missing
+        ([9] * 11, True, 0),
+    ])
+def test_splitting_categorical_no_splits(X_binned, has_missing_values,
+                                         n_bins_non_missing):
+    # Checks categorical splits are correct when there are no spliits
+
+    n_bins = max(X_binned) + 1
+    n_samples = len(X_binned)
+    X_binned = np.array([X_binned], dtype=X_BINNED_DTYPE).T
+    X_binned = np.asfortranarray(X_binned)
+
+    l2_regularization = 0.0
+    min_hessian_to_split = 1e-3
+    min_samples_leaf = 1
+    min_gain_to_split = 0.0
+
+    sample_indices = np.arange(n_samples, dtype=np.uint32)
+    all_gradients = np.ones(n_samples, dtype=G_H_DTYPE)
+    has_missing_values = np.array([False], dtype=np.uint8)
+    all_hessians = np.ones(1, dtype=G_H_DTYPE)
+    sum_gradients = all_gradients.sum()
+    sum_hessians = n_samples
+    hessians_are_constant = True
+
+    builder = HistogramBuilder(X_binned, n_bins, all_gradients,
+                               all_hessians, hessians_are_constant)
+
+    n_bins_non_missing = np.array([n_bins_non_missing], dtype=np.uint32)
+    monotonic_cst = np.array([MonotonicConstraint.NO_CST] * X_binned.shape[1],
+                             dtype=np.int8)
+    is_categorical = np.ones_like(monotonic_cst, dtype=np.uint8)
+    missing_values_bin_idx = n_bins - 1
+
+    splitter = Splitter(X_binned, n_bins_non_missing,
+                        missing_values_bin_idx, has_missing_values,
+                        is_categorical, monotonic_cst,
+                        l2_regularization, min_hessian_to_split,
+                        min_samples_leaf, min_gain_to_split,
+                        hessians_are_constant)
+
+    histograms = builder.compute_histograms_brute(sample_indices)
+    value = compute_node_value(sum_gradients, sum_hessians,
+                               -np.inf, np.inf, l2_regularization)
+    split_info = splitter.find_node_split(n_samples, histograms,
+                                          sum_gradients, sum_hessians, value)
+
+    # no split found
+    assert split_info.gain == -1
+
+
+def _assert_threshold_equals_bitset(expected_thresholds, bitset):
+    # bitset is assumed to be an array 8 of uint32
+
+    # form bitset from threshold
+    expected_threshold_bitset = np.zeros(8, dtype=np.uint32)
+    for threshold in expected_thresholds:
+        i1 = threshold // 32
+        i2 = threshold % 32
+        expected_threshold_bitset[i1] |= 1 << i2
+
+    # check for equality
+    assert_array_equal(expected_threshold_bitset, bitset)
+
+
+@pytest.mark.parametrize(
+    "X_binned, all_gradients, expected_thresholds, n_bins_non_missing,"
```

Comment:
we should not use "threshold" when talking about categories.

Should this be "expected_categories_left" or something similar?